lib/chai/core/assertions.js

```diff
@@ -504,10 +499,42 @@ module.exports = function (chai, _) {
       return;
     }
 
-    // Assert inclusion in an array or substring in a string.
+    var isDeep = flag(this, 'deep')
+      , descriptor = isDeep ? 'deep ' : ''
+      , included = false;
+
+    switch (objType) {
+      case 'string':
+        included = obj.indexOf(val) !== -1;
+        break;
+
+      case 'weakmap':
+      case 'weakset':
+        if (isDeep) {
+          var flagMsg = flag(this, 'message')
+            , ssfi = flag(this, 'ssfi');
+          flagMsg = flagMsg ? flagMsg + ': ' : '';
+
+          throw new AssertionError(
+            flagMsg + 'unable to use .deep.include with weak collection',
+            undefined,
+            ssfi
+          );
+        }
+
+        included = obj.has(val);
+        break;
+
+      default:
+        var isEql = isDeep ? _.eql : function(a, b) { return a === b; };
+        obj.forEach(function (item) {
+          included = included || isEql(item, val);
+        });
```

Comment:
We can optimize away looping through `map` and `set` for non-deep `include` (cool) and use `some` + `indexOf` on arrays. Code:

```diff
+  case 'map':
+  case 'set':
+    if (isDeep) {
+      obj.forEach(function (item) {
+        included = included || _.eql(item, val);
+      });
+   } else {
+     included = obj.has(val);
+   }
+   break;
+
+  case 'array':
+    if (isDeep) {
+     included = obj.some(function (item) {
+       return _.eql(item, val);
+     })
+    } else {
+      included = obj.indexOf(val) !== -1;
+    }
+    break;

-  default:
-    var isEql = isDeep ? _.eql : function(a, b) { return a === b; };
-    obj.forEach(function (item) {
-      included = included || isEql(item, val);
-    });
```